streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBTimeOrderedKeyValueBufferTest.java

```diff
@@ -0,0 +1,168 @@
+        streamsMetrics = new StreamsMetricsImpl(metrics, "test-client", StreamsConfig.METRICS_LATEST, new MockTime());
+        context = new MockInternalNewProcessorContext<>(StreamsTestUtils.getStreamsConfig(), new TaskId(0, 0), TestUtils.tempDirectory());
+    }
+
+    public void createJoin(final Duration grace) {
+        final RocksDBTimeOrderedKeyValueBytesStore store = new RocksDbTimeOrderedKeyValueBytesStoreSupplier("testing",  100).get();
+        buffer = new RocksDBTimeOrderedKeyValueBuffer<>(store, grace, "testing");
+        buffer.setSerdesIfNull(serdeGetter);
+        store.init((StateStoreContext) context, store);
+        buffer.init((StateStoreContext) context, store);
+    }
+
+    private void pipeRecord(final String key, final String value, final long time) {
+        final Record<String, String> record = new Record<>(key, value, time);
+        context.setRecordContext(new ProcessorRecordContext(time, offset++, 0, "testing", new RecordHeaders()));
+        buffer.put(time, record, context.recordContext());
+    }
+
+    @Test
+    public void shouldAddAndEvictRecord() {
+        createJoin(Duration.ZERO);
+        final AtomicInteger count = new AtomicInteger(0);
+        pipeRecord("1", "0", 0L);
+        buffer.evictWhile(() -> buffer.numRecords() > 0, r -> count.getAndIncrement());
+        assertThat(count.get(), equalTo(1));
+    }
+
+    @Test
+    public void shouldAddAndEvictRecordTwice() {
+        createJoin(Duration.ZERO);
+        final AtomicInteger count = new AtomicInteger(0);
+        pipeRecord("1", "0", 0L);
+        buffer.evictWhile(() -> buffer.numRecords() > 0, r -> count.getAndIncrement());
+        assertThat(count.get(), equalTo(1));
+        pipeRecord("2", "0", 1L);
+        buffer.evictWhile(() -> buffer.numRecords() > 0, r -> count.getAndIncrement());
+        assertThat(count.get(), equalTo(2));
+    }
+
+    @Test
+    public void shouldAddAndEvictRecordTwiceWithNonZeroGrace() {
+        createJoin(Duration.ofMillis(1));
+        final AtomicInteger count = new AtomicInteger(0);
+        pipeRecord("1", "0", 0L);
+        buffer.evictWhile(() -> buffer.numRecords() > 0, r -> count.getAndIncrement());
+        assertThat(count.get(), equalTo(0));
+        pipeRecord("2", "0", 1L);
+        buffer.evictWhile(() -> buffer.numRecords() > 0, r -> count.getAndIncrement());
+        assertThat(count.get(), equalTo(1));
+    }
+
+    @Test
+    public void shouldAddRecordsTwiceAndEvictRecordsOnce() {
+        createJoin(Duration.ZERO);
+        final AtomicInteger count = new AtomicInteger(0);
+        pipeRecord("1", "0", 0L);
+        buffer.evictWhile(() -> buffer.numRecords() > 1, r -> count.getAndIncrement());
+        assertThat(count.get(), equalTo(0));
+        pipeRecord("2", "0", 1L);
+        buffer.evictWhile(() -> buffer.numRecords() > 0, r -> count.getAndIncrement());
+        assertThat(count.get(), equalTo(2));
+    }
+
+    @Test
+    public void shouldDropLateRecords() {
```

Comment:
I think you do not need to use `evictWhile()` for this test. You can just verify if `numRecords()` does not increase after calling `put()`. The same is true for the next test.

Reply:
that works